src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/RandomMouseObjectActionData.cs

```diff
@@ -0,0 +1,258 @@
+                if (!preconditionsMet)
+                {
+                    preconditionsMet = currentUITransforms.Any(a => a.Value.screenSpaceBounds != null && StateRecorderUtils.OptimizedContainsStringInList(preconditionNormalizedPaths, a.Value.NormalizedPath));
+                }
+
+                if (!preconditionsMet)
+                {
+                    preconditionsMet = currentGameObjectTransforms.Any(a => a.Value.screenSpaceBounds != null && StateRecorderUtils.OptimizedContainsStringInList(preconditionNormalizedPaths, a.Value.NormalizedPath));
+                }
+
+                if (!preconditionsMet)
+                {
+                    //TODO: Someday make this only log the ones that weren't matched instead of all the preconditions
+                    StringBuilder error = new StringBuilder(500);
+                    error.Append("RandomMouseClicker - Missing one or more precondition normalized paths\r\n");
+                    var preconditionNormalizedPathsCount = preconditionNormalizedPaths.Count;
+                    for (var i = 0; i < preconditionNormalizedPathsCount; i++)
+                    {
+                        var pc = preconditionNormalizedPaths[i];
+                        error.Append(pc);
+                        if (i + 1 < preconditionNormalizedPathsCount)
+                        {
+                            error.Append("\r\n");
+                        }
+                    }
+
+                    return error.ToString();
+                }
+
+                if (currentGameObjectTransforms.Count == 0 || uiOrGameObject && currentUITransforms.Count > 0)
+                {
+                    possibleTransformsToClick = currentUITransforms.Values.Where(a => a.screenSpaceBounds != null).ToList();
+                }
+                else
+                {
+                    possibleTransformsToClick = currentGameObjectTransforms.Values.Where(a => a.screenSpaceBounds != null).ToList();
+                }
+
+                var possibleTransformsCount = possibleTransformsToClick.Count;
+                if (possibleTransformsCount > 0)
+                {
+                    var RESTART_LIMIT = 20;
+                    var restartCount = 0;
+                    // try up to 20 times per frame to find a valid click object, but after that, give up till the next frame as none may be available
+                    // helps prevent long searches, or cases where the user excludes the whole screen
+                    do
+                    {
+                        var transformOption = possibleTransformsToClick[Random.Range(0, possibleTransformsCount)];
+                        if (StateRecorderUtils.OptimizedStringStartsWithStringInList(excludedNormalizedPaths, transformOption.NormalizedPath))
+                        {
+                            // not allowed object .. pick another
+                            continue; // while
+                        }
```

Comment:
Is the idea here that sampling is faster than an upfront filter?

Reply:
If you pick one, and evaluate a small handle of exclusion rules; that is far more efficient than evaluating 100s of options through the exclusion rules when you're still only going to pick one.  There is a limit of up to 20 tries per update to find a 'good' object, so in theory we could miss.  I was fine with that as it is a random clicker and not clicking for 1 frame seems fine